File: src/schemas/assembly.ts
import { z } from 'zod';
import { CompositionSchema } from './composition';

export const MOST_TILES = 256;
export const MOST_VERTICES = 8192;
export const MOST_SOURCES = 12;

/** `assembly_weight::FEATHER`: a recipe's feather where it names none. */
export const DEFAULT_FEATHER = 0.0025;
/** The widest feather the page offers. */
export const MOST_FEATHER = 0.015;
/** The widest a stored recipe may name, past the page's own: the page clamps what it reads. */
const MOST_STORED_FEATHER = 0.1;

/**
 * What a tile asks the frame it picks for.
 *
 * `subject` reads that frame where the thing under the reader's click went, so a person who moved
 * is taken as they stand there. `ground` reads it in place, so what comes back is whatever stands
 * there instead - which is how a thing is removed rather than replaced.
 */
export const TakesSchema = z.enum(['subject', 'ground']);
export type Takes = z.infer<typeof TakesSchema>;

/** `[a, b, c, d, tx, ty]` of `[a b; c d] * p + t` over the canvas: where a source is read for `p`. */
const WarpSchema = z.tuple([z.number(), z.number(), z.number(), z.number(), z.number(), z.number()]);
export type Warp = z.infer<typeof WarpSchema>;

export const SeamsSchema = z.object({
  pick: z.array(z.number().int().nonnegative()).max(MOST_TILES),
  // What each tile asked for when these were solved, which a render holds against the recipe's own:
  // a tile flipped between its subject and the ground names the same frame and means another shape.
  takes: z.array(TakesSchema).max(MOST_TILES).optional(),
  base: z.number().int().nonnegative(),
  vertices: z.array(z.tuple([z.number(), z.number()])).max(MOST_VERTICES),
  tiles: z.array(z.array(z.number().int().nonnegative()).min(3)).max(MOST_TILES),
  source: z.array(z.number().int().nonnegative()).max(MOST_TILES),
  // The tile each piece grew from, which a click on it opens.
  zone: z.array(z.number().int().nonnegative()).max(MOST_TILES),
  // The room each piece's seam has, as a share of the long edge: what §5.2's feather is sized from.
  corridor: z.array(z.number()).max(MOST_TILES),
  // A piece's frame read where its content went (`patch_search::tracked`), or in place for a tile
  // asking for the ground.
  warp: z.array(WarpSchema).max(MOST_TILES),
  // A piece's gain to meet what lies across its seams.
  exposure: z.array(z.number().positive()).max(MOST_TILES),
});
export type Seams = z.infer<typeof SeamsSchema>;

// The panorama's geometry, plus the tiles across it. Bounded because `photos.recipe` is one
// replicated cell re-parsed by a SQL trigger on every write to the row.
export const AssemblyRecipeSchema = CompositionSchema.extend({
  vertices: z.array(z.tuple([z.number(), z.number()])).max(MOST_VERTICES),
  tiles: z.array(z.array(z.number().int().nonnegative()).min(3)).max(MOST_TILES),
  pick: z.array(z.number().int().nonnegative()),
  base: z.number().int().nonnegative(),
  // What each tile asks its pick for: the subject the reader pointed at, wherever that frame shows
  // it, or the ground that frame shows in its place. Absent, every tile asks for the subject, which
  // is what every recipe written before removal existed meant.
  takes: z.array(TakesSchema).max(MOST_TILES).optional(),
  // The widest §5.2's feather may be, as a share of the long edge: the reader's to set.
  feather: z.number().min(0).max(MOST_STORED_FEATHER).optional(),
  // Where the picked frames actually meet, solved for `pick` and `base` over the carve's volume
  // (`assembly_seams`). Absent, the tiles are the seams.
  seams: SeamsSchema.optional(),
  // The layer key the carve's seam volume was written under, which a re-solve reads.
  seamVolume: z.string().regex(/^[A-Za-z0-9_-]+$/).optional(),
}).superRefine((recipe, ctx) => {
  const bad = (message: string): void => void ctx.addIssue({ code: 'custom', message });
  const tiles = recipe.tiles.length;
  if (recipe.sources.length > MOST_SOURCES) bad(`an assembly is made of at most ${MOST_SOURCES} photographs`);
  if (recipe.pick.length !== tiles) bad('pick has to have one entry a tile');
  if (recipe.base >= recipe.sources.length) bad('the base names no source');
  if (recipe.pick.some((source) => source >= recipe.sources.length)) bad('a pick names no source');
  if (recipe.takes != null && recipe.takes.length !== tiles) bad('takes has to have one entry a tile');
  if (recipe.tiles.some((loop) => loop.some((at) => at >= recipe.vertices.length))) bad('a tile names no vertex');
  const seams = recipe.seams;
  if (seams != null) {
    const pieces = seams.tiles.length;
    for (const [name, field] of [['source', seams.source], ['zone', seams.zone], ['corridor', seams.corridor], ['warp', seams.warp], ['exposure', seams.exposure]] as const) {
      if (field.length !== pieces) bad(`the seams have to have one ${name} a piece`);
    }
    if (seams.source.some((source) => source >= recipe.sources.length)) bad('a piece names no source');
    if (seams.zone.some((zone) => zone >= tiles)) bad('a piece names no tile');
    if (seams.tiles.some((loop) => loop.some((at) => at >= seams.vertices.length))) bad('a piece names no vertex');
  }
});
export type AssemblyRecipe = z.infer<typeof AssemblyRecipeSchema>;

/** A seam solve for each of `picks` in place of the recipe's own: one a frame of an open tile. */
export const SeamsRequestSchema = z
  .object({
    recipe: AssemblyRecipeSchema,
    picks: z.array(z.array(z.number().int().nonnegative()).max(MOST_TILES)).min(1).max(MOST_SOURCES + 2),
  })
  .superRefine(({ recipe, picks }, ctx) => {
    if (picks.some((pick) => pick.length !== recipe.tiles.length)) {
      ctx.addIssue({ code: 'custom', message: 'a pick set has to have one entry a tile' });
    }
    if (picks.some((pick) => pick.some((source) => source >= recipe.sources.length))) {
      ctx.addIssue({ code: 'custom', message: 'a pick names no source' });
    }
  });

export const SolvedSeamsSchema = z.array(SeamsSchema.nullable());

/** §4.2's settled preview: one recipe, rendered as its picks stand. */
export const PreviewRequestSchema = z.object({ recipe: AssemblyRecipeSchema });

export const CommitAssemblyRequestSchema = z.object({ recipe: AssemblyRecipeSchema });

/**
 * What the analysis answers: an untiled recipe the reader seeds tiles on, and whether §3.1's corner
 * check held.
 *
 * Parsed rather than passed through, for `AlignedSchema`'s reason: a reply this build cannot read
 * is a refusal here rather than a page drawing tiles out of undefined.
 */
export const AnalysedSchema = z.object({
  recipe: AssemblyRecipeSchema,
  // §3.1's corner check failed: the frames were not taken from one place.
  unaligned: z.boolean(),
  warnings: z.array(z.string()),
});
export type Analysed = z.infer<typeof AnalysedSchema>;

export const SolvedSeamsResponseSchema = z.object({ seams: SolvedSeamsSchema.nullable() });
export type SolvedSeamsResponse = z.infer<typeof SolvedSeamsResponseSchema>;

export const AssemblyPreviewSchema = z.object({ url: z.string() });
export type AssemblyPreview = z.infer<typeof AssemblyPreviewSchema>;

export const AssemblyJobStartedSchema = z.object({ jobId: z.string() });
export type AssemblyJobStarted = z.infer<typeof AssemblyJobStartedSchema>;

/**
 * What a carve answers: what the analysis measured, and §4.3's layers.
 *
 * The layers are URLs rather than bytes, one per `analysed.recipe.sources` in that order - they are
 * pictures, and the page decodes each through `ImageDecoder` as it does every other rendition.
 */
export const CarvedSchema = z.object({
  analysed: AnalysedSchema,
  layers: z.array(z.string()),
});
export type Carved = z.infer<typeof CarvedSchema>;

/** An analysis `startAssembly` began, as it stands now. */
export const AssemblyJobSchema = z.object({
  id: z.string(),
  // The frames being analysed, which the page names its swatches from before the analysis answers.
  photoIds: z.array(z.string()),
  status: z.enum(['analysing', 'ready', 'failed', 'cancelled']),
  // How much of the analysis is behind it, 0 to 1.
  fraction: z.number(),
  // Once `ready`.
  carved: CarvedSchema.optional(),
  // Once `failed`.
  error: z.string().optional(),
});
export type AssemblyJob = z.infer<typeof AssemblyJobSchema>;

/** What §2.7's reopen answers: the recipe as it was committed, and the same layers over it. */
export const ReopenedAssemblySchema = z.object({
  recipe: AssemblyRecipeSchema,
  // One per `recipe.sources`, or empty where a source is missing: a canvas is drawn from every frame
  // it names or from none.
  layers: z.array(z.string()),
  // Sources since deleted or binned, which is what opens the page read-only.
  missingSources: z.array(z.string()),
});
export type ReopenedAssembly = z.infer<typeof ReopenedAssemblySchema>;

File: native/rawshim/src/assembly.rs
//! What an assembly is: a composition's geometry, and the tiles across it that each take one frame.
//! A recipe, not a picture, for [`crate::composition`]'s reasons.

use crate::composition::Composition;
use crate::light::Gain;
use crate::px::Share;
use serde::{Deserialize, Serialize};

/// What a tile asks the frame it picks for.
///
/// **The whole of the difference between replacing a thing and removing one.** A tile taking
/// [`Takes::Subject`] reads its frame where the thing under the reader's click went, so a person
/// who moved between frames is taken as they stand there; one taking [`Takes::Ground`] reads it
/// in place, so what comes back is whatever stands there instead.
#[derive(Serialize, Deserialize, Clone, Copy, Debug, Default, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum Takes {
    #[default]
    Subject,
    Ground,
}

/// A composition with tiles across it, each of which takes its pixels from one source.
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Assembly {
    #[serde(flatten)]
    pub spec: Composition,
    /// Canvas pixels the tile loops are drawn from.
    pub vertices: Vec<[f32; 2]>,
    /// One loop of vertex indices a tile: the least of what its pick takes (`assembly_seams`).
    pub tiles: Vec<Vec<u32>>,
    /// Which source each tile takes.
    pub pick: Vec<usize>,
    /// What each tile asks that source for. Shorter than [`Assembly::tiles`], or absent, where the
    /// recipe was written before a tile could ask for anything but its subject.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub takes: Vec<Takes>,
    /// The source every pixel outside every tile comes from. One more pick, not a fallback: every
    /// tile has one.
    pub base: usize,
    /// The most §5.2's `W(x)` may be anywhere, as a share of the long edge: the reader's.
    #[serde(default = "default_feather")]
    pub feather: f32,
    /// Where the picked frames actually meet, solved for `pick` and `base` (`assembly_seams`).
    /// Absent, the tiles are the seams.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub seams: Option<crate::assembly_seams::Seams>,
}

fn default_feather() -> f32 {
    crate::assembly_weight::FEATHER.raw() as f32
}

impl Assembly {
    /// A recipe over `spec` with no tiles yet, every pixel taking its reference.
    pub fn untiled(spec: Composition) -> Assembly {
        Assembly {
            base: spec.reference,
            vertices: Vec::new(),
            tiles: Vec::new(),
            pick: Vec::new(),
            takes: Vec::new(),
            feather: default_feather(),
            seams: None,
            spec,
        }
    }

    /// The recipe as a render draws it: the solved pieces, each under its own warp and balance, or
    /// the tiles themselves where nothing was solved.
    pub fn rendered(&self) -> Result<Drawing, String> {
        let drawing = |vertices, tiles, pick, corridor, warp, gain| Drawing {
            spec: self.spec.clone(),
            vertices,
            tiles,
            pick,
            base: self.base,
            corridor,
            warp,
            gain,
            feather: self.feather,
        };
        let Some(seams) = &self.seams else {
            let tiles = self.tiles.len();
            // A tile has no corridor of its own, so its blend is the narrowest the render draws.
            return Ok(drawing(
                self.vertices.clone(),
                self.tiles.clone(),
                self.pick.clone(),
                vec![0.0; tiles],
                vec![crate::composition::no_warp().map(|at| at as f32); tiles],
                vec![Gain::ONE; tiles],
            ));
        };
        // The takes as well as the picks: a tile flipped between its subject and the ground names
        // the same frame, and the seams that follow are a different shape entirely. Compared tile
        // by tile, since a list left empty and one naming every subject mean the same recipe.
        let takes_differ = (0..self.tiles.len()).any(|tile| {
            let solved = seams.takes.get(tile).copied().unwrap_or_default();
            solved != self.takes(tile)
        });
        if seams.pick != self.pick || seams.base != self.base || takes_differ {
            return Err("the seams were solved for other picks than the recipe's".into());
        }
        Ok(drawing(
            seams.vertices.clone(),
            seams.tiles.clone(),
            seams.source.clone(),
            seams.corridor.clone(),
            seams.warp.clone(),
            seams
                .exposure
                .iter()
                .map(|&gain| Gain::of_ratio(f64::from(gain)))
                .collect(),
        ))
    }

    /// [`Assembly::feather`] as the share it is.
    pub fn feather(&self) -> Share {
        Share::measured(f64::from(self.feather), 1.0)
    }

    /// What tile `tile` asks its pick for.
    pub fn takes(&self, tile: usize) -> Takes {
        self.takes.get(tile).copied().unwrap_or_default()
    }
}

/// An assembly as a render draws it: pieces, each taking one frame through its own warp and gain.
#[derive(Clone, Debug)]
pub struct Drawing {
    pub spec: Composition,
    /// Canvas pixels the piece loops are drawn from.
    pub vertices: Vec<[f32; 2]>,
    pub tiles: Vec<Vec<u32>>,
    /// Which source each piece takes.
    pub pick: Vec<usize>,
    /// The source every pixel outside every piece comes from.
    pub base: usize,
    /// Per piece, the room its seam has, **as a share of the long edge**, capped at the piece's own
    /// inradius so that §5.2's feather cannot be wider than the piece is deep. What `W(x)` is built
    /// from.
    pub corridor: Vec<f32>,
    /// Per piece, the affine over the **canvas** that says where its frame is read -
    /// `[a, b, c, d, tx, ty]` of `[a b; c d] * p + t` - so that what it shows lands where the base
    /// showed it.
    pub warp: Vec<[f32; 6]>,
    /// Per piece, what its frame's light is multiplied by to meet what lies across its seams.
    pub gain: Vec<Gain>,
    /// As [`Assembly::feather`].
    pub feather: f32,
}

impl Drawing {
    /// [`Drawing::feather`] as the share it is.
    pub fn feather(&self) -> Share {
        Share::measured(f64::from(self.feather), 1.0)
    }

    /// Piece `tile`'s warp, in `f64`.
    pub fn warp_of(&self, tile: usize) -> [f64; 6] {
        self.warp[tile].map(f64::from)
    }

    /// The base first, then every other distinct pick ascending.
    ///
    /// The order is the render's: §5.2's mask strides by it, so a second ordering anywhere is a
    /// picture taking the wrong frame.
    pub fn sources_used(&self) -> Vec<usize> {
        let mut rest: Vec<usize> = self
            .pick
            .iter()
            .copied()
            .filter(|&p| p != self.base)
            .collect();
        rest.sort_unstable();
        rest.dedup();
        std::iter::once(self.base).chain(rest).collect()
    }
}
